### PR TITLE
Add layout printers to the debugger and autogenerate `tools/debug_printer`

### PR DESCRIPTION
**Pull request by antalsz**
This improves the UX for working with the debugger

### CODE REVIEW

File: tools/Makefile [hunk 1]
```diff
@@ -328,6 +328,12 @@ endif
 clean::
 	rm -f -- caml-tex caml-tex.exe caml_tex.cm?
 
+# Debug printer script
+debug_printers: debug_printers.ml debug_printers.cmo
+	echo 'load_printer "tools/$(basename $<).cmo"' > '$@'
+	awk '{ print "install_printer Debug_printers." $$2 }' \
```
Comment (goldfirere) [inline tools/Makefile:334]: I don't know `awk`, but my interpretation from this line is that it's slurping the lines from `debug_printers.ml` and looking at the second word of each line. This means that an editor of `debug_printers.ml` now needs to be careful to use exactly the same formatting as previous lines. In addition, if there is not an existing printer that suits someone's needs, they will not be able to add extra facilities here. Maybe that's all OK, but a loud comment in `debug_printers.ml` would seem to save a future person lots of head-scratching.

File: typing/layouts.ml [hunk 2]
```diff
@@ -84,6 +84,22 @@ module Sort = struct
   and equate s1 s2 = match s1 with
     | Const c1 -> equate_sort_const s2 c1
     | Var v1 -> equate_var v1 s2
+
+  module Debug_printers = struct
+    open Format
+
+    let rec t ppf = function
+      | Var v   -> fprintf ppf "Var %a" var v
+      | Const c -> fprintf ppf (match c with
```
Comment (goldfirere) [inline typing/layouts.ml:93]: I would turn that match into a `string_of_const` function elsewhere I think.

**Review by goldfirere — COMMENTED** [review]
I'm very happy to have more debug printers installed. I'm a little less convinced about the automation to refresh `debug_printers`: it constrains the way we write `debug_printers.ml` in a way that may prove annoying and counterintuitive later, for very little savings. And when things break, fixing it requires knowledge of `awk`. (Maybe that's widespread?)

In particular, I believe @ncik-roberts recently added something to `debug_printers.ml` that does not conform to this specification (in a branch). Is that accurate? But maybe it's better to put the printing code right in the compiler anyway? I'm really on the fence here.

File: typing/layouts.ml (hunk 2 above)
Comment (ncik-roberts) [reply typing/layouts.ml:93]: I don't think that the debug printer I added will be problematic for Antal's change. It follows the format of other printers we list in `debug_printers.ml`.

**Comment by ncik-roberts** [comment]
I don't think that the debug printer I added will be problematic for Antal's change. It follows the format of other printers we list in `debug_printers.ml`.

File: typing/layouts.ml (hunk 2 above)
Comment (antalsz) [reply typing/layouts.ml:93]: I've found that leaving them as printers is more convenient for consistency, but I'll make it a separate `let const ppf = ...` function above

Comment (antalsz) [reply typing/layouts.ml:93]: Oh heh and in doing that I fixed a small printing bug :-)

File: tools/Makefile (hunk 1 above)
Comment (antalsz) [reply tools/Makefile:334]: Comment added; as discussed with you, there's already an `awk` dependency in the testsuite stuff, so we're not adding new requirements

**Comment by antalsz** [comment]
I improved this by causing the automation to look only for lines that start with `"let "`, and added a comment to `tools/debug_printers.ml`, which per conversation with @goldfirere should address the concerns

**Review by goldfirere — APPROVED** [review]
After merging, please put a PSA to the type systems group that this behavior has changed.

Looks good!

**Comment by antalsz** [comment]
I'll put out the PSA after ocaml-flambda#148 merges :+1:

**Pull request merged**